ddtrace/tracer/span.go

```diff
@@ -518,14 +520,12 @@ func (s *span) String() string {
 		fmt.Sprintf("Type: %s", s.Type),
 		"Tags:",
 	}
-	s.RLock()
 	for key, val := range s.Meta {
 		lines = append(lines, fmt.Sprintf("\t%s:%s", key, val))
 	}
 	for key, val := range s.Metrics {
 		lines = append(lines, fmt.Sprintf("\t%s:%f", key, val))
 	}
-	s.RUnlock()
 	return strings.Join(lines, "\n")
```

Comment:
Do you have benchmarks on the impact of this? We now have the `strings.Join` inside the lock, and it shouldn't need to be. In general, let's try to keep the `*span` locked for the absolute minimum amount of time. A potentially less impactful way to make this change would be to do something like this:

```
func (s *span) String() string {
	defer s.RUnlock()
	lines := []string{
		fmt.Sprintf("Name: %s", s.Name),
		fmt.Sprintf("Service: %s", s.Service),
		fmt.Sprintf("Resource: %s", s.Resource),
		fmt.Sprintf("TraceID: %d", s.TraceID),
		fmt.Sprintf("SpanID: %d", s.SpanID),
		fmt.Sprintf("ParentID: %d", s.ParentID),
		fmt.Sprintf("Start: %s", time.Unix(0, s.Start)),
		fmt.Sprintf("Error: %d", s.Error),
		fmt.Sprintf("Type: %s", s.Type),
		"Tags:",
	}
        // Hold lock only for span types that 
	s.RLock()
        lockedLines := []string{ fmt.Sprintf("Duration: %s", time.Duration(s.Duration)) }
	for key, val := range s.Meta {
		lockedLines = append(lockedLines, fmt.Sprintf("\t%s:%s", key, val))
	}
	for key, val := range s.Metrics {
		lockedLines = append(lockedLines, fmt.Sprintf("\t%s:%f", key, val))
	}
        s.RUnlock()
        lines = append(lines, lockedLines...)
	return strings.Join(lines, "\n")
```

Reply:
Sorry I am a bit confused, here I followed the recommendation from @knusbaum. I thought in this case `for good measure` meant we wanted to be simple and safe here.
 
I am wondering if there is a use case to call `String()` for performance here of it would be just debug informations. However, you are right that's why my initial change only moved the lock https://github.com/DataDog/dd-trace-go/pull/1337/commits/d6c2ea6c514c8c0aa95640593967c490d28005b9.

The proposed change above does not fix the data race issue, we could either keep the defer with something like:
```
func (s *span) String() string {
	s.RLock()
	func() { 
		defer s.RUnlock()
		// lines code ...
	}()
	return strings.Join(lines, "\n")
}
```
(^ I am not a fan of it)
or just keep my initial change only and don't defer.

Also, I wouldn't unlock twice this will have other side effects.

Let me know which solution you prefere. It does seems it would add a lot of complexity for the value here, but this is my personal opinion.

Reply:
@katiehockman which use case were you thinking of where this method would be critical for performance? As far as I know it is only used for debugging.

Reply:
Gotcha, didn't realize this was just used for debugging and won't affect customers. Will resolve.